[IMP] base, web, product: `KanbanView`, adjust `<aside>` sizing

Brieuc-brd opened the pull request:
### Adjust `<aside>` sizing 
PR odoo/https://github.com/odoo/odoo/pull/167751, introduced Bootstrap columns to manage kanban cards using an `aside` element.
But the rendering is not optimal for some cards (eg. Contacts, Products) and the these are less easy to read.

This commit adjusts the default size of `<aside>` and its inner image elements. Bootstrap columns can still be defined if we want to customize the size of the `<main>` and `<aside>`.

### Reintroduce `min-width` on `<main>`
PR odoo/https://github.com/odoo/odoo/pull/167751, also removed a `min-width` defined on the `<main>` that was useful to force its content to be truncated. The property has been reintroduced since it's still necessary.

task-3997115



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/169790.png)](https://mergebot.odoo.com/odoo/odoo/pull/169790)

Brieuc-brd commented:
Hey @stefanorigano 👋 
Could you take a look at this PR ?

stefanorigano commented on addons/web/static/src/views/kanban/kanban_controller.scss:
```diff
@@ -103,8 +106,17 @@
         }
 
+        aside {
+            flex: 1 0 var(--KanbanAside-width);
```
Why an `<aside/>` should be  allowed to grow outside the width defined by `--KanbanAside-width` 🤔 ?
I think that `<main/>` should be `flex-grow: 1` and this line should probably be something like 
```suggestion
            flex-basis: var(--KanbanAside-width);
```
could you check?

stefanorigano reviewed (changes requested):
Hello @Brieuc-brd , thanks for your work :)
Here some remarks for you , thanks again

Brieuc-brd commented on addons/web/static/src/views/kanban/kanban_controller.scss:
```diff
@@ -39,6 +39,11 @@
     --KanbanColumn__highlight-background: #{tint-color($o-action, 90%)};
     --KanbanColumn__highlight-border: #{$o-action};
 
+    --KanbanAside-width: 95px;
```
Good catch! I'll use them instead 👍 .

> Also, those values are still used to set --KanbanRecord__image-width and  --KanbanRecord__image--fill-width, but apparent these are both unused now(?) 😕 Could you check?

They're used in `kanban_record_legacy.scss`. 🙂

Brieuc-brd commented on addons/product/views/product_template_views.xml:
```diff
@@ -95,8 +95,8 @@
                 <templates>
-                    <t t-name="kanban-card" class="row g-0">
-                        <main class="col-10 pe-2">
+                    <t t-name="kanban-card" class="o_product_kanban_card flex-row">
```
This should be necessary. A Kanban card is supposed to be `flex-column` by default.
When removing [this](https://github.com/odoo/odoo/pull/169790/commits/6350f7b7f0e3a5b68b1c573f47a47c2c8a1fb910#diff-368e56feea5bb220fcafabbf45b40131fdd3d496fe0f2643a9df5cdec61cb190L88-L90), I forgot to adapt correctly the `flex-column`. It's fixed ✅

Brieuc-brd commented:
Thanks for your feedbacks @stefanorigano 🙏 
I've updated the branch, could you take a look at it ?

stefanorigano commented on addons/product/views/product_template_views.xml:
```diff
@@ -114,8 +114,8 @@
                         </main>
-                        <aside class="col-2 p-0">
-                            <field name="image_128" widget="image" alt="Product"/>
+                        <aside class="flex-shrink-0">
```
Why `flex-shrink-0` is needed? I think that's just because  of `w-100`  applied on `<main>`, but it looks unnecessary as well :)

stefanorigano replied:
> If flex-shrink-0 is not defined, the width could be reduced according to the main content even without w-100.

.. meaning "_`aside` is broken unless we use this class in each template_" .. If it's the case it's better to add the rule in CSS instead 😉

In a previous comment I've proposed to remove `grow = 1` and  test flex-basis  ( https://github.com/odoo/odoo/pull/169790#discussion_r1645972279) ... I think you applied my suggestion without testing it and probably applied `shrink-0`after to overcome the undesired results.
  
  If `shrink` was still necessary you could have keep it rather than replacing it with utility classes 😉

Brieuc-brd replied:
Ok I see your point 👍.
I've removed these utility classes in favor of global css.

stefanorigano commented on addons/product/views/product_template_views.xml:
```diff
@@ -95,8 +95,8 @@
-                    <t t-name="kanban-card" class="row g-0">
-                        <main class="col-10 pe-2">
+                    <t t-name="kanban-card" class="o_product_kanban_card flex-row">
+                        <main class="flex-grow-1 w-100 pe-2">
```
I don't see a scenario in which `<main>` shouldn't grow, IMO we can add `flex-grow: 1;`  in the CSS directly and remove the class form here.

stefanorigano commented on odoo/addons/base/views/res_partner_views.xml:
```diff
@@ -218,11 +218,11 @@
-                                                <field name="avatar_128" widget="image" options="{'img_class': 'object-fit-cover'}" alt="Contact image"/>
+                                        <t t-name="kanban-card" class="flex-row">
+                                            <aside class="o_kanban_aside_full d-flex flex-shrink-0">
+                                                <field name="avatar_128" class="align-self-center w-100" widget="image" options="{'img_class': 'object-fit-cover'}" alt="Contact image"/>
                                             </aside>
-                                            <main class="col-9 col-sm-10 col-md-9 ps-2">
+                                            <main class="flex-grow-1 w-100 ps-2 ps-md-0">
```
Here as well,  like in the "products" template. I'd suggest:
- `<main>`: remove `flex-grow-1 w-100`
- `<aside>`:  remove `flex-shrink-0`
- css: unsure that `main` is `flex-grow: 1` globally

ps: tested briefly in the browser, seems like this setup makes `d-flex`  and `align-self-center` classes unnecessary for `aside` and its field. Could you check?

Brieuc-brd replied:
> `<aside>` : remove `flex-shrink-0`

Same as [the last comment](https://github.com/odoo/odoo/pull/169790#discussion_r1652747861), I don't think we should remove it :).

> ps: tested briefly in the browser, seems like this setup makes d-flex and align-self-center classes unnecessary for aside and its field. Could you check?

I've made some tests, these classes are necessary to keep the vertical alignment of `aside`.

stefanorigano replied:
> Same as https://github.com/odoo/odoo/pull/169790#discussion_r1652747861, I don't think we should remove it :).

Replied there  :) 

> these classes are necessary to keep the vertical alignment of aside.

Not on mobile ;)
These rules seems needed when "aside full" is active only... if confirmed  we can added them in the css directly.  

| with `d-flex` and `align-self-center` | without |
|--------|--------|
| ![image](https://github.com/odoo/odoo/assets/8558351/7552c558-57f7-453b-bba5-e53bf93dd8dc) | ![image](https://github.com/odoo/odoo/assets/8558351/f9343c98-e162-4b0b-9840-568215f69100) |

Brieuc-brd replied:
> Not on mobile ;)
These rules seems needed when "aside full" is active only... if confirmed we can added them in the css directly.

I've kept it to achieve the same result as before the introduction of `cols` in cards.
But as discussed, centering the image doesn't really make sense. So I've totally removed the centered alignment.
What do you think ?

stefanorigano reviewed (changes requested):
Hello @Brieuc-brd ,  here some remarks for you.
Thanks :)

Brieuc-brd commented:
Thanks for your review @stefanorigano 👍 
I've updated the branch and replied to some of your comments.

Could you take a look at it ?

Brieuc-brd commented:
Thanks again for your review @stefanorigano 🙏 
I've updated the branch with the last changes, could you have a look at it ?
Thanks in advance !

The diff : https://github.com/odoo/odoo/compare/e5dc0af0d0c6afab03af3d4da10035b4f414970d..55eb8eabbb686d2d88989ac0a09e247cc59b4d2e

stefanorigano commented on addons/web/static/src/views/kanban/kanban_controller.scss:
```diff
@@ -103,13 +103,29 @@
+            &.o_kanban_aside_full {
+                --Card-Aside-Image-height: 100%;
+
+                flex: 0 0 var(--KanbanRecord__image--fill-width);
```
What do you think? :)
```suggestion
                --KanbanRecord__image-width: var(--KanbanRecord__image--fill-width);
```

stefanorigano commented on addons/web/static/src/views/kanban/kanban_controller.scss:
```diff
@@ -103,13 +103,29 @@
+            flex: 0 0 var(--KanbanRecord__image-width);
+
+            img {
+                height: var(--Card-Aside-Image-height, var(--KanbanRecord__image-width));
```
OK, basically we're using `--KanbanRecord` prefix everywhere except for the footer's gap  🤔

I thought that `--KanbanRecord`  was legacy and `--Card` was used for the "new design"... but apparently it's not the case.
I was partially wrong [here](https://github.com/odoo/odoo/pull/169790#discussion_r1645446245), sorry for the headache 🙂

Still, `-KanbanAsideImage-height` isn't a good fit neither, we should replicate the current variable naming IMO:

```
--KanbanRecord__image-width    <- exists already
--KanbanRecord__image-height   <- new
```

What do you think?

stefanorigano reviewed (changes requested):
Thanks @Brieuc-brd , courage, almost there 💪
Some nitpicking comments for you, plus:
- commit's title: there is  enough room to list all the affected modules + I don't think that "fixed" it's the right term, since values are suggested and can be easily adjusted.  I'd suggest:
`[IMP] base, web, product: 'KanbanView', adjust '<aside>' sizing`
- Commit/pr description: we should probably write something about main's `min-width: 0` and the fact that the layout it's `flex-flow: column;` by default now.


Thanks!

Brieuc-brd commented:
Thanks @stefanorigano 👍 
I've updated the branch, the commit title/description and the PR title/description.

> ... and the fact that the layout it's flex-flow: column; by default now.

As it's already been done [here](https://github.com/odoo/odoo/pull/167751/commits/89954730117fef5f5766cd6735a6debf3357171f), I've just added an explanation for the `min-width` implementation.

The diff : https://github.com/odoo/odoo/compare/05d7c6fe3f54d4db803a4331ba1bcb624752100a..4619d04ceec42bf684f2ea46133ec05712d0e581

Could you take a look?

stefanorigano reviewed (approved):
LGTM, thanks @Brieuc-brd 🙌

Brieuc-brd commented:
Hey @aab-odoo 👋 
This PR is ready for tech. review, could you have a look at it ?
Thanks in advance 🙂

aab-odoo reviewed (approved):
robodoo r+

Brieuc-brd commented:
Hey @aab-odoo 👋 

Sorry to interrupt the merge, but I’ve spotted an issue :
The default style defined in this PR prevents the use of Bootstrap columns, there was a conflict with css priorities.
I've fixed the issue.

Also, the `max-height` of the direct child of `o_kanban_aside_full` was set in all cases when it should be set when the `<aside>` uses an image. To fix this, I've created a custom class to be applied where necessary.

I've also updated the commit/PR message.

The diff : https://github.com/odoo/odoo/compare/eea0b9d71f36094631d599fcb71da93a751211e7..77ccfc714388c88d25a185ba45eda75a3d48a46e

Could you have a look ?
Thanks in advance 🙏

aab-odoo commented:
robodoo r+
